### ISSUE TITLE
Proposal:  Test Case DisplayName Extensibility

### ISSUE THREAD

**Issue by NickCraver** [xunit/xunit#1429]
This is a proposal with ideas given some cases I've hit on larger projects. I'll use `StackExchange.Redis.Tests` as my example library here.

## Background

What I have today is something like this for a unit test name:
```
StackExchange.Redis.Tests.Booksleeve.Issues.<method>
```
**All** tests start with the assembly namespace, so the "StackExchange.Redis.Tests" is all noise and repeated. This creates truncation issues, removing more relevant data in most lists. For example, Appveyor:
<img width="631" alt="screen shot 2017-08-27 at 08 38 33" src="https://user-images.githubusercontent.com/454813/29751126-51f92972-8b18-11e7-9e57-eb2b71d6a238.png">

What I want is to remove that `DisplayName` starting string in all cases. At first, I tried a local-to-my-namespace `[Fact]` attribute, like this:
```c#
public class FactAttribute : Xunit.FactAttribute
{
    public override string DisplayName
    {
        get => base.DisplayName?.Replace("StackExchange.Redis.Tests.", "");
        set => base.DisplayName = value;
    }
}
```
...but that doesn't work. Since the way XUnit works overall is that `DisplayName` is null on the attribute (makes sense) and *if null* then uses the format specified. Currently [the options for this](https://xunit.github.io/docs/configuring-with-json.html) are "method" or "classAndMethod" (fully qualified including namespace), but those are the only 2 options.

So then I tried overriding the discoverer and setting `XUnitTestCase.DisplayName`, but this is `protected set`. So without reflection, that doesn't work either...you have to override the whole chain down to test case generation and that gets a bit nuts. I'm currently [using the reflection method](https://github.com/StackExchange/StackExchange.Redis/pull/700/commits/a4ff89dc76bdb954a4961cb5f264b86369b0afd5) as this seemed the simplest thing to do, for example:

```c#
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Reflection;
using Xunit.Abstractions;
using Xunit.Sdk;

namespace StackExchange.Redis.Tests
{
    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
    [XunitTestCaseDiscoverer("StackExchange.Redis.Tests.FactDiscoverer", "StackExchange.Redis.Tests")]
    public class FactAttribute : Xunit.FactAttribute { }

    public class FactDiscoverer : Xunit.Sdk.FactDiscoverer
    {
        public FactDiscoverer(IMessageSink diagnosticMessageSink) : base(diagnosticMessageSink) { }

        public override IEnumerable<IXunitTestCase> Discover(ITestFrameworkDiscoveryOptions discoveryOptions, ITestMethod testMethod, IAttributeInfo factAttribute)
        {
            foreach (var test in base.Discover(discoveryOptions, testMethod, factAttribute))
            {
                yield return test.SetName(dn => dn?.Replace("StackExchange.Redis.Tests.", ""));
            }
        }
    }

    internal static class XUnitExtensions
    {
        private static readonly PropertyInfo dnProtected =
            typeof(XunitTestCase).GetProperty(nameof(XunitTestCase.DisplayName), BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);

        public static IXunitTestCase SetName(this IXunitTestCase testCase, Func<string, string> replace)
        {
            if (testCase is XunitTestCase xutc)
            {
                dnProtected.SetValue(xutc, replace(xutc.DisplayName));
            }
            return testCase;
        }
    }
}
```
<sub>(note: bits truncated, e.g. duplication for Theory and error handling)</sub>

So unless I missed a path, it's just generally hard to tweak `DisplayName` on a test case today.

## Proposal

What if we added a new property to the runner configuration beside `methodDisplay`. For example `methodDisplayFormat` which takes a string. Optionally, this could require setting a `methodDisplay` of `custom`. 

Here's an example JSON config reflecting the default:
```json
{
  "methodDisplayFormat": "{namespace}.{class}.{method}"
}
```
This would make things more extensible more easily for many people, but it doesn't cover the case I have. More tokens could take care of that though, let's say we have a method in assembly `StackExchange.Redis.Tests`, today this is presented as:
```
StackExchange.Redis.Tests.Booksleeve.Issues.Issue10.TestA
```
The tokens available would be:

| Token | Value |
| ------ | ------ |
| `namespace` |"StackExchange.Redis.Tests.Booksleeve.Issues" |
| `namespaceWithoutAssembly` | "Booksleeve.Issues" |
| `class` | "Issue10" |
| `method` | "TestA" |

...and perhaps some other tokens I can't think of?

My current code could be replaced by:
```json
{
  "methodDisplayFormat": "{namespaceWithoutAssembly}.{class}.{method}"
}
```

### Open Questions:
- Is this a terrible idea?
- Should generic arguments live in the string, or just be appended?
  - Same question for `[Theory]` arguments
- What other tokens would be useful?
- What use cases does this not cover?

### Side proposal
I think for the more programmatic users who want to control this is crazy ways, we could simply [remove `protected` from `TestMethodTestCase`](https://github.com/xunit/xunit/blob/bbcb06af9a836041407729989f467aa4cf638ad4/src/xunit.execution/Sdk/Frameworks/TestMethodTestCase.cs#L71), which would remove the need for full inheritance or (more likely?) reflection workarounds.

Thoughts? I just wanted to write this up and see what other peoples' use cases were and if this would help or greatly misses the mark.

**Comment by saitolabs** [comment]
+1

**Comment by Tasteful** [comment]
+1

Why not 'simpleClassAndMethod' that will ignore all namespace parts that are same for all test cases?

**Comment by el2iot2** [comment]
+1. Immediately useful to me. Although tasteful's would work too.

**Comment by NickCraver** [comment]
@Tasteful that's certainly another route, but ultimately less generally extensible. For example, that's not really intuitive IMO, it would imply (to me) `<class>.<method>` with no namespace parts at all. When looking at a solution, which "shared" pieces are removed? Is it per project? Is it per solution? (not really how the runner works)

I think we can answer all of those things, it's just not an intuitive thing, at least to me. Or we could simply change up the name: `withoutBaseNamespace` or some such.

 I think having the format string could just replace the current enum altogether long-term, while being more flexible at the same time. I also think more tokens are likely useful, though I can't think of them right now. They tell you pretty clearly what the format would be and you can change them up however you want. Users wouldn't have to wait or lobby for additional format combinations to be added to an enum (so the same change allows a greater range of scenarios), they could just combine the tokens however they wanted. They'd still have to file issues/PRs for additional tokens though, granted.

**Comment by marcind** [comment]
I frequently nest test classes within other classes (the outer class providing shared scope/context logic):

```csharp
class ComponentTests {
  class InScenario1 : Component1Tests {
    [Fact] ItWorks() { Assert.True(true); }
  }
}
```

I guess the outer class name becomes part of the inner class's namespace, but maybe a token to represent the distinction would be helpful?

**Comment by Tasteful** [comment]
@NickCraver no. That route is not so extensible that a format settings. If we look at the aspnet dev guidelines https://github.com/aspnet/Home/wiki/Engineering-guidelines#unit-test-class-naming the test-class should live in the same namespace as the tested class, that will mean that assembly name and namespaces not will match.

The csproj use the 'RootNamespace' as propertyname for the 'namespace without assemblyname'. Maybe that can be added as an assembly attribute and used to remove first parts of the names.

Another option is that 'methodDisplay' will take the enum or a factory-class that is used for generate the names, then it will be verry flexible and no specifik tokens need to be created.

**Comment by NickCraver** [comment]
@Tasteful I wouldn't go by *any* conventions here, usage is all over the map I agree so guidelines from anyone about how classes *should* be arranged gos by the wayside fast :)

I agree configuring a class *in code* (as all others are, usually via an attribute - likely assembly level) also makes sense, with the format string being the default implementation of such an extensibility point. I don't see these features as mutually exclusive but rather complementary and in the spirit of xUnit extensibility in general.

**Comment by bradwilson** [comment]
@marcind wrote:

> I guess the outer class name becomes part of the inner class's namespace, but maybe a token to represent the distinction would be helpful?

Not really, no. The class name is `Outer+Inner`; it's not part of the namespace, nor would we treat it like one.

**Comment by bradwilson** [comment]
@Tasteful wrote:

> Another option is that 'methodDisplay' will take the enum or a factory-class that is used for generate the names, then it will be verry flexible and no specifik tokens need to be created.

I like this a lot better than the token scenario; and in fact, the token scenario could be implemented by way of something like this. What would be missing is the ability to look at the existing configuration values.

**Comment by drewburlingame** [comment]
I have a need for this as well, and additionally, I need to be able to skip some Theory data and have them shown as skipped.  Currently, I just have to remove them from the list of tests and they become hidden.    I would also like to provide traits so I can mark records with Issue# and categories.

I think NUnit has a good solution for this with [TestCaseData](https://github.com/nunit/docs/wiki/TestCaseData).  Adopting this concept would provide a way to extend Theory data with a way to provide Name, Skipped, Traits, etc.

**Comment by Sergey-Terekhin** [comment]
Hi everyone!
Any news on this?

**Comment by bradwilson** [comment]
No, and for reasons I can't say at the moment, this never made it onto my v3 roadmap, so I haven't been thinking about it. I'll put it on the v3 roadmap now to make sure it's something that gets considered.
